include/mbedtls/legacy_or_psa.h

```diff
@@ -209,4 +209,88 @@
 #define MBEDTLS_HAS_ALG_SHA_512_VIA_MD_OR_PSA_BASED_ON_USE_PSA
 #endif
 
+/* Ciphers and cipher chaining using low-level or PSA based on availability */
+#if defined(MBEDTLS_AES_C) || \
+    ( defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_KEY_TYPE_AES) )
+#define MBEDTLS_HAS_ALG_AES_VIA_LOWLEVEL_OR_PSA
+#endif
+
+#if defined(MBEDTLS_ARIA_C) || \
+    ( defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_KEY_TYPE_ARIA) )
+#define MBEDTLS_HAS_ALG_ARIA_VIA_LOWLEVEL_OR_PSA
+#endif
+
+#if defined(MBEDTLS_CAMELLIA_C) || \
+    ( defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_KEY_TYPE_CAMELLIA) )
+#define MBEDTLS_HAS_ALG_CAMELLIA_VIA_LOWLEVEL_OR_PSA
+#endif
+
+#if defined(MBEDTLS_CIPHER_MODE_CBC) || \
+    ( defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_ALG_CBC_NO_PADDING) )
+#define MBEDTLS_HAS_ALG_CBC_VIA_LOWLEVEL_OR_PSA
+#endif
+
+/* Ciphers using Mbed TLS ciphers or PSA based on availability */
+#if ( defined(MBEDTLS_CIPHER_C) && defined(MBEDTLS_AES_C) ) || \
+    ( !defined(MBEDTLS_CIPHER_C) && \
+        defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_KEY_TYPE_AES) )
+#define MBEDTLS_HAS_ALG_AES_VIA_CIPHER_OR_PSA
+#endif
+
+#if ( defined(MBEDTLS_CIPHER_C) && defined(MBEDTLS_ARIA_C) ) || \
+    ( !defined(MBEDTLS_CIPHER_C) && \
+        defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_KEY_TYPE_ARIA) )
+#define MBEDTLS_HAS_ALG_ARIA_VIA_CIPHER_OR_PSA
+#endif
+
+#if ( defined(MBEDTLS_CIPHER_C) && defined(MBEDTLS_CAMELLIA_C) ) || \
+    ( !defined(MBEDTLS_CIPHER_C) && \
+        defined(MBEDTLS_PSA_CRYPTO_C) && defined(PSA_WANT_KEY_TYPE_CAMELLIA) )
+#define MBEDTLS_HAS_ALG_CAMELLIA_VIA_CIPHER_OR_PSA
+#endif
+
+/* Ciphers, chaining and modes using MbedTLS ciphers or PSA
+ * based on MBEDTLS_USE_PSA_CRYPTO */
+#if ( !defined(MBEDTLS_USE_PSA_CRYPTO) && \
+        defined(MBEDTLS_CIPHER_C) && defined(MBEDTLS_AES_C) ) || \
+    ( defined(MBEDTLS_USE_PSA_CRYPTO) && defined(PSA_WANT_KEY_TYPE_AES) )
+#define MBEDTLS_HAS_ALG_AES_VIA_CIPHER_OR_PSA_BASED_ON_USE_PSA
```

Comment:
As discussed in chat, this is not a good interface because it's expressed in terms of how it's implemented, rather than in terms of how to use it.

I think this really does correspond do a useful interface: “if I'm calling TLS, is this cryptographic mechanism supported?”. The name should reflect this.